os/fs/fat/fs_fat32.c

```diff
@@ -0,0 +1,2678 @@
+static int fat_readdir(FAR struct inode *mountpt, FAR struct fs_dirent_s *dir);
+static int fat_rewinddir(FAR struct inode *mountpt, FAR struct fs_dirent_s *dir);
+
+static int fat_bind(FAR struct inode *blkdriver, FAR const void *data, FAR void **handle);
+static int fat_unbind(FAR void *handle, FAR struct inode **blkdriver, unsigned int flags);
+static int fat_statfs(FAR struct inode *mountpt, FAR struct statfs *buf);
+
+static int fat_unlink(FAR struct inode *mountpt, FAR const char *relpath);
+static int fat_mkdir(FAR struct inode *mountpt, FAR const char *relpath, mode_t mode);
+static int fat_rmdir(FAR struct inode *mountpt, FAR const char *relpath);
+static int fat_rename(FAR struct inode *mountpt, FAR const char *oldrelpath, FAR const char *newrelpath);
+static int fat_stat_common(FAR struct fat_mountpt_s *fs, FAR uint8_t *direntry, FAR struct stat *buf);
+static int fat_stat_file(FAR struct fat_mountpt_s *fs, FAR uint8_t *direntry, FAR struct stat *buf);
+static int fat_stat_root(FAR struct fat_mountpt_s *fs, FAR struct stat *buf);
+static int fat_stat(struct inode *mountpt, const char *relpath, FAR struct stat *buf);
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/* See fs_mount.c -- this structure is explicitly extern'ed there.
+ * We use the old-fashioned kind of initializers so that this will compile
+ * with any compiler.
+ */
+
+const struct mountpt_operations fat_operations = {
+	fat_open,					/* open */
+	fat_close,					/* close */
+	fat_read,					/* read */
+	fat_write,					/* write */
+	fat_seek,					/* seek */
+	fat_ioctl,					/* ioctl */
+
+	fat_sync,					/* sync */
+	fat_dup,					/* dup */
+	fat_fstat,					/* fstat */
+	fat_truncate,				/* truncate */
+
+	fat_opendir,				/* opendir */
+	NULL,						/* closedir */
+	fat_readdir,				/* readdir */
+	fat_rewinddir,				/* rewinddir */
+
+	fat_bind,					/* bind */
+	fat_unbind,					/* unbind */
+	fat_statfs,					/* statfs */
+
+	fat_unlink,					/* unlinke */
+	fat_mkdir,					/* mkdir */
+	fat_rmdir,					/* rmdir */
+	fat_rename,					/* rename */
+	fat_stat					/* stat */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: fat_open
+ ****************************************************************************/
+
+static int fat_open(FAR struct file *filep, FAR const char *relpath, int oflags, mode_t mode)
+{
+	struct fat_dirinfo_s dirinfo;
+	FAR struct inode *inode;
+	FAR struct fat_mountpt_s *fs;
+	FAR struct fat_file_s *ff;
+	uint8_t *direntry;
+	int ret;
+
+	/* Sanity checks */
+
+	DEBUGASSERT(filep->f_priv == NULL && filep->f_inode != NULL);
```

Comment:
Also check for "filep" variable.

Reply:
is it necessary to check for it because if filep is NULL then its members will be NULL .

Reply:
Yes, it might result in assertion.

Reply:
Ok